enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,366 @@
+
+1. How do we handle validation which is done today in go? (ovnk)
+   Would it make sense to have a validation.yaml that includes shell script
+   to be run on the system and validate anything necessary?, or should
+   we simply stop doing this type of validation?
+
+2. How do we handle CR upgrades when moving to newer versions of the APIs?,
+   See the `Upgrading`section.
+
+### Test Plan
+
+We will run additional test matrix with `microshift-core` alone, in those
+tests any OVN/NetworkPolicy tests would need to be disabled.
+
+Additional unit tests should cover the loading of core components as
+kustomize manifests, as well as the depencency handing between them.
+
+Additional unit tests should cover the loading of multiple `.d` style
+directories with manifests.
+
+## Implementation History
+
+* [openshift/microshift](https://github.com/openshift/microshift)
+* [Design guidelines](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+## Alternatives
+
+### Using go plugins which microshift can load when available, intead of yamls.
+
+[go/plugin](https://pkg.go.dev/plugin) could be used to create an interface
+for loadable components, MicroShift would scan an spefic directory for .so files
+
+Every plugin would contain it's own list of images, setup functionality,
+validation hooks, etc.
+
+The Risks here are:
+
+* Binary code duplication: How do we handle shared code between MicroShift
+  and the plugins, for example plugins would be expected to call kustomize
+  libraries, but can't call back into code in MicroShift binary. This would
+  force us replicate those libraries again statically in the plugins along
+  with the kubernetes clients
+
+* The inteface and calling convention needs to be maintained with versioning
+  to avoid externally created plugins from breaking as we evolve the API.
+  This is also true for yaml based components, but it's simpler to provide
+  additive iterations without breaking backwards compatibility in that case.
+  
+* CGO_ENABLED needs to be enabled.
+
+* Requires same version of go vendor dependencies.
+
+The possible benefits are:
+
+* Easier to handle validations of the system
+* More flexible ability to read and handle configuration files.
+* Upgrade of CRDs can potentially be done
+
+### on/off components within MicroShift binary code
+
+We could add flags to the MicroShift configuration which then can control
+the component loading via flags.
+
+Image versions would be switched at compile time for upstream versions
+where necessary. Components which we could not ship upstream would be disabled
+out from compilation via build -tags.
+
+All the code relevant to the non-loaded componets would remain in the binary.
+
+The possible benefits are:
+
+* It's probably the simplest implementation.
+
+Risks:
+
+* It does not prevent coupling creep. Components would continue to couple with
+  the MicroShift codebase in unexpected ways, as we already see today.
+
+### External binaries
```

Comment:
What I had envisioned for this was to have microshift run a per-component script that would generate some manifests on startup, and then microshift would load those manifests in order. 

That allows components to use kustomize if they want, but it also allows them to do something more complex (or simpler) if that makes sense.

It also means that those "binaries" do not need kube clients (listed as a risk for this approach). They could be shell scripts, in some cases.

Reply:
Hmm, that could be another approach yes. Having binaries/(or shell scripts) generate the kustomize scripts (or just yamls) for you. In this way we also solve the limitation of no-templating in kustomize, we could pass down config details (like ClusterIP ranges, pod IP ranges, etc... as environment variables. If a plugin needs to become more complex for some reason they could become binaries)

I actually like the idea.

Let's see what others think in tomorrow's meeting.

We were converging around the idea that we could have a mixed approach where we allowed kustomize scripts side by side with an external binary for system checks or other system tasks... but this is likely to be more flexible, we'd just need to define the API contract between MicroShift and the scripts.

Reply:
If the plugins need config values from MicroShift, they can use the command to get MicroShift to emit its current configuration. That way we don't have to keep maintaining a set of variables or command line arguments.

Reply:
`get MicroShift to emit its current configuration`, smart!
I can see how binary embed the assets and generate them in the known kustomize directory.
With script, how the assets are delivered? via the rpm package or as plain text in the shell script?

Reply:
Do we see value adding the binary/script hooks both before and after the kustomize? the latter one may be used for validating the readiness of plugin (not the pod readiness), since kustomize won't return errors besides a format issue.

Reply:
This approach would allow to get rid of microshift-ovs-init, right?
On start up plugin could configure host level networking and emit the yamls, so MicroShift can take them and apply without giving plugin the kubeconfig.

Related to kubeconfig - do we want to prohibit plugin from just looking for it on the disk? This might be more of a hardening discussion, but should we jail/chroot/isolate the plugins (like to prohibit from using the konfig), but then how could they make changes to configure host level networking? Because rn we would run them as root...

As for contract and communication with plugin: we could take a page from CNI binaries book: 
- pass json via stdin (although that would add a requirement of jq on the system because parsing json with pure bash is not my cup of tea),
- pass args via envs 
- expect a json as stdout (creating output json by hand shouldn't be too cumbersome).

Reply:
Output json can contain some instruction for MicroShift about what to do first with the yamls - maybe `kustomize . | oc delete -f -` before `kustomize . | oc create -f -` (re. upgrades)?
Plugin would need to somehow get the information about current (if any) version of the plugin running in the cluster but that could be it's responsibility to create a file like `components.d/10-plug/.metadata` or smth like this.

Not sure how long we'd go with bash (shell? I wouldn't except anything other than shell or Go be used here) plugins as some helpers would be most welcome to reduce the burden on devs and reduce bug possibility

Reply:
@pmtk  microshift-ovs-init.service can be moved to anywhere right before applying the ovnk manifests. like you said, it may require root to touch host networking configurations.  I don't see there is a strong need for this script/binary to access kubeconfig, besides applying manifests which is moved to use kustomize. maybe as a development guidance, the binary/script is not supposed to read the kubeconfig on the disk.

About the contract and communication, the CNI approach doesn't expect any stdout (log messages) from script/binary, otherwise would be considerred as CNI json data.

Reply:
I'm not clear what the diff is between this alternative and the current proposal. It seems that:

- Both assume each component plug-in is responsible to have a script or binary that generates manifests which MicroShift then loads.
- It may be useful to allow "simpler or more complex" mechanisms than kustomize. I can see that, but would argue the same would hold true for our current way of allowing users to add workloads ("manifest auto-applier"). If so, it'd make sense to extend the "manifest auto-applier" from kustomize-only to also apply normal manifests (if no kustomize.yaml is present) and to use a "helm template" run as more standard way of doing a "go-templating". That pattern is called helm-kustomize.
- We need _some_ way of providing config to CNI plug-ins. It's been mentioned plug-ins could call `microshift show-config` to do plug-in specific rendering in the script/binary. Or - if we extended the "manifest auto-applier" towards a helm-kustomize pattern, MicroShift could render a `values.yaml` (Helm's standard way of parameterizing charts) for example.
- The diff seems to be this proposal assumes MicroShift calls the script/binary with a new API we have to define, while the current proposal assumes plug-in authors are responsible to create a systemd unit that starts the script/binary before/after MicroShift starts. I'm not clear what the use cases are where adding a new API would be preferable over systemd dependencies?

Reply:
> I'm not clear what the diff is between this alternative and the current proposal. It seems that:
> 
>     * Both assume each component plug-in is responsible to have a script or binary that generates manifests which MicroShift then loads.
> 
>     * It may be useful to allow "simpler or more complex" mechanisms than kustomize. I can see that, but would argue the same would hold true for our current way of allowing users to add workloads ("manifest auto-applier" / "kustomizer service"). If so, it'd make sense to extend this "manifest auto-applier" from kustomize-only to also apply normal manifests (if no kustomize.yaml is present) and to use a "helm template" run as more standard way of doing a "go-templating" that also includes conditional logic/loops if really necessary. That pattern is called helm-kustomize. This way we maintain the (IMO useful) property that this new plug-in mechanism and our current "user plug-in mechanism" remain one and the same.

Maybe microshift can execute the plugin script/binary only when exits and support a fallback option to render/load manifests directly from yaml template, the fallback option is the "simpler" way.
> 
>     * We need _some_ way of providing config to CNI plug-ins. It's been mentioned plug-ins could call `microshift show-config` to do plug-in specific rendering in the script/binary. Or - if we extended the "manifest auto-applier" towards a helm-kustomize pattern, MicroShift could render a `values.yaml` (Helm's standard way of parameterizing charts) for example.
> 
>     * The diff seems to be this proposal assumes MicroShift calls the script/binary with a new API we have to define, while the current proposal assumes plug-in authors are responsible to create a systemd unit that starts the script/binary before/after MicroShift starts. I'm not clear what the use cases are where systemd units are not sufficient but we need to define a new API that works in parallel to the CNI/CSI APIs?

The API definition between microshift and plugin seems not necessary at this point, the script/binary can be called w/o arguments and output the rendered manifests to stdout.

Reply:
IMHO having systemd files pre-generate config files on boot for MicroShift to consume afterwards is going to be problematic, i.e. microshift show-config does not show the node IP detected by microshift:

```
[root@rhel9 ~]# microshift show-config
apiServer:
  subjectAltNames:
  - rhel9.host-only
  - rhel9.local
  - rhel9.shared
debugging:
  logLevel: Normal
dns:
  baseDomain: example.com
network:
  clusterNetwork:
  - cidr: 10.42.0.0/16
  serviceNetwork:
  - 10.43.0.0/16
  serviceNodePortRange: 30000-32767
node:
  hostnameOverride: ""
  nodeIP: ""
```

For example, if we needed to incorporate anything that requests an update from the plugin (i.e. host IP changed or  other event).


@zshi-redhat I believe some level of API is necessary (at least for the node IP).

Reply:
@mangelajo using `microshift show-config --mode=effective` would have shown that node IP.